src/lambda/commands/uploadLambda.ts

```diff
@@ -0,0 +1,246 @@
+    window = Window.vscode()
+): Promise<telemetry.Result> {
+    const workspaceFolders = vscode.workspace.workspaceFolders || []
+
+    const parentDirArr = await window.showOpenDialog({
+        canSelectFolders: true,
+        canSelectFiles: false,
+        canSelectMany: false,
+        defaultUri: workspaceFolders[0]?.uri,
+    })
+
+    if (!parentDirArr || parentDirArr.length !== 1) {
+        return 'Cancelled'
+    }
+
+    return await zipAndUploadDirectory(functionNode, parentDirArr[0].fsPath)
+}
+
+async function runUploadLambdaZipFile(
+    functionNode: LambdaFunctionNode,
+    window = Window.vscode()
+): Promise<telemetry.Result> {
+    const workspaceFolders = vscode.workspace.workspaceFolders || []
+
+    const zipFileArr = await window.showOpenDialog({
+        canSelectFolders: false,
+        canSelectFiles: true,
+        canSelectMany: false,
+        defaultUri: workspaceFolders[0]?.uri,
+        filters: {
+            'ZIP archive': ['zip'],
+        },
+    })
+
+    if (!zipFileArr || zipFileArr.length !== 1) {
+        return 'Cancelled'
+    }
+
+    const zipFile = fs.readFileSync(zipFileArr[0].fsPath)
+
+    return await uploadZipBuffer(functionNode, zipFile)
+}
+
+async function runUploadLambdaWithSamBuild(
+    functionNode: LambdaFunctionNode,
+    window = Window.vscode()
+): Promise<telemetry.Result> {
+    const workspaceFolders = vscode.workspace.workspaceFolders || []
+
+    const parentDirArr = await window.showOpenDialog({
+        canSelectFolders: true,
+        canSelectFiles: false,
+        canSelectMany: false,
+        defaultUri: workspaceFolders[0]?.uri,
+    })
+
+    if (!parentDirArr || parentDirArr.length !== 1) {
+        return 'Cancelled'
+    }
+
+    // TODO: Detect if handler is present and error out prematurely?
+    // TODO: What should we do if no manifest (package.json, requirements.txt, etc.) is present? Directory won't be `sam build`-able.
+
+    try {
+        const invoker = getSamCliContext().invoker
+        const parentDir = parentDirArr[0]
+
+        const tempDir = await makeTemporaryToolkitFolder()
+        ExtensionDisposableFiles.getInstance().addFolder(tempDir)
+        const templatePath = path.join(tempDir, 'template.yaml')
+
+        // TODO: Use an existing template file if it's present?
+
+        await new SamTemplateGenerator()
+            .withFunctionHandler(functionNode.configuration.Handler!)
+            .withResourceName('tempResource')
+            .withRuntime(functionNode.configuration.Runtime!)
+            .withCodeUri(parentDir.fsPath)
+            .generate(templatePath)
+
+        const buildDir = path.join(tempDir, 'output')
+        await new SamCliBuildInvocation({
+            buildDir,
+            templatePath,
+            invoker,
+            skipPullImage: true,
+            useContainer: false,
+            baseDir: parentDir.fsPath,
+        }).execute()
+
+        // TODO: Remove template file if generating one?
```

Comment:
+1. Is there any reason not to remove it?

Reply:
Looks like this handles best if you use the folder it builds to that's adjacent to the template. Will require more manual testing but we don't need an explicit removal.